### PR TITLE
Fix MaintenanceSolution restores from remote folders

### LINKED ISSUE #3013: Restore-DbaDatabase fails when using OlaH's maintenance solution backups
## Bug Report

I'm attempting to copy a database named BBLEARN over a database named BBLEARN_TEST located on the same SQL Server using Restore-DbaDatabase, and receiving a warning from Get-OlaHRestoreFile (captured here with -Verbose output included):

```
VERBOSE: [Restore-DbaDatabase][13:16:54] Changing statement timeout to infinity
VERBOSE: [Restore-DbaDatabase][13:16:54] ParameterSet  = Restore
VERBOSE: [Restore-DbaDatabase][13:16:54] Unverified input, full scans - H:\Backups\DEVSQL02\BBLEARN\FULL\
VERBOSE: [Get-DbaBackupInformation][13:16:54] Not using sql for H:\Backups\DEVSQL02\BBLEARN\FULL\
VERBOSE: [Get-OlaHRestoreFile][13:16:54] Starting
VERBOSE: [Get-OlaHRestoreFile][13:16:54] Checking Path
WARNING: [Get-OlaHRestoreFile][13:16:54] H:\Backups\DEVSQL02\BBLEARN\FULL\ is not valid
VERBOSE: [Restore-DbaDatabase][13:16:54] Processing DatabaseName - BBLEARN_TEST
VERBOSE: [Format-DbaBackupInformation][13:16:54] Starting
VERBOSE: [Format-DbaBackupInformation][13:16:54] String passed in for DB rename
VERBOSE: [Restore-DbaDatabase][13:16:54] VerifyOnly = False
VERBOSE: [Get-DbaDbPhysicalFile][13:16:54] Connecting to DEVSQL02
VERBOSE: [Restore-DbaDatabase][13:16:54]  passed testing
VERBOSE: [Restore-DbaDatabase][13:16:54] Passing in to restore
```
The SQL Server in question backs up all its databases using the Ola Hallengren maintenance solution, and I've confirmed that the path "H:\Backups\DEVSQL02\BBLEARN\FULL\" is valid using Test-DbaSqlPath:

image

Output of Get-DbatoolsLog:

```
1/10/2018 1:16:53 PM Restore-DbaDatabase         InternalComment                      Starting
1/10/2018 1:16:53 PM Restore-DbaDatabase         Debug                                Parameters bound: MaintenanceSolutionBackup, SqlInstance, Path, WhatIf, WithReplace, Verbose, DestinationLogDirectory, ReplaceDbNameI...
1/10/2018 1:16:54 PM Restore-DbaDatabase         Verbose                              Changing statement timeout to infinity
1/10/2018 1:16:54 PM Restore-DbaDatabase         Verbose                              ParameterSet  = Restore
1/10/2018 1:16:54 PM Restore-DbaDatabase         Verbose                              Unverified input, full scans - H:\Backups\DEVSQL02\BBLEARN\FULL\
1/10/2018 1:16:54 PM Get-DbaBackupInformation    InternalComment                      Starting
1/10/2018 1:16:54 PM Get-DbaBackupInformation    Debug                                Parameters bound: SqlInstance, DirectoryRecurse, MaintenanceSolution, IgnoreLogBackup
1/10/2018 1:16:54 PM Get-DbaBackupInformation    VeryVerbose                          Not using sql for H:\Backups\DEVSQL02\BBLEARN\FULL\
1/10/2018 1:16:54 PM Get-OlaHRestoreFile         Verbose                              Starting
1/10/2018 1:16:54 PM Get-OlaHRestoreFile         Verbose                              Checking Path
1/10/2018 1:16:54 PM Get-OlaHRestoreFile         Warning                              H:\Backups\DEVSQL02\BBLEARN\FULL\ is not valid
1/10/2018 1:16:54 PM Restore-DbaDatabase         Verbose                              Processing DatabaseName - BBLEARN_TEST
1/10/2018 1:16:54 PM Format-DbaBackupInformation Verbose                              Starting
1/10/2018 1:16:54 PM Format-DbaBackupInformation Verbose                              String passed in for DB rename
1/10/2018 1:16:54 PM Restore-DbaDatabase         Verbose                              VerifyOnly = False
1/10/2018 1:16:54 PM Get-DbaDbPhysicalFile       Verbose                              Connecting to DEVSQL02
1/10/2018 1:16:54 PM Get-DbaDbPhysicalFile       Debug                                SELECT DB_NAME(database_id) AS Name, physical_name AS PhysicalName FROM sys.master_files
1/10/2018 1:16:54 PM Restore-DbaDatabase         Verbose                               passed testing
1/10/2018 1:16:54 PM Restore-DbaDatabase         Verbose                              Passing in to restore
```

### General Troubleshooting steps

- [x] Verified running the latest release of dbatools? (Running 0.9.153)
- [x] Verified errors are not related to permissions? (Powershell running under account that has the Sysadmin role on the SQL Server)
- [x] Can duplicate in new/clean PowerShell session (clean = `powershell -NoProfile`)?

### Version Information
 - Operating System (Name|Version): Windows 10 Enterprise version 1703
 - PowerShell Version: 5.1.15063.726
 - SQL Server (Edition|Version): SQL Server 2014 Enterprise SP1 (12.0.4232)

## Steps to Reproduce

This code returns $True:

```powershell
$SqlServer = "DEVSQL02"
$SourceDatabase = "BBLEARN"
$PathToSourceDatabaseBackups = "H:\Backups\$SqlServer\$SourceDatabase\FULL\"

Test-DbaSqlPath -SqlInstance $SqlServer -Path $PathToSourceDatabaseBackups
```
And here's my complete restore code, which is throwing the warning:

```powershell
$SqlServer = "DEVSQL02"
$SourceDatabase = "BBLEARN"
$PathToSourceDatabaseBackups = "H:\Backups\$SqlServer\$SourceDatabase\FULL\"

$RestoreParams = @{
    SqlInstance = $SqlServer
    DatabaseName = "$($SourceDatabase)_TEST"
    Path = $PathToSourceDatabaseBackups
    WithReplace = $true
    MaintenanceSolutionBackup = $true
    IgnoreLogBackup = $true
    ReplaceDbNameInFile = $true
    DestinationDataDirectory = "E:\Data\$DestinationDatabase"
    DestinationLogDirectory = "F:\Logs\$DestinationDatabase"
    WhatIf = $true
    Verbose = $false

}
Restore-DbaDatabase @RestoreParams
```

Please let me know if there's any additional details I can provide.

### PR DESCRIPTION
**Pull request by Stuart-Moore**
## Type of Change

 - [x] Bug fix (non-breaking change, fixes #3013 )
 - [ ] New feature (non-breaking change, adds functionality)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 

### Purpose
Issue #3031 appears to be caused by OlaStyle restores using GCI, which they've been doing for about a year so suprising we've not come across this before!

Have rewritten things to that they also work with XpDirTree and still keep the IgnoreLogBackups thing working.

Get-OlaHRestoreFile has been removed as it's no longer required.

Passes pester happily and works like I think it should do, but needs some more testing as it's a reasonable change,.

### CODE REVIEW

**Comment by Zhunya** [comment]
Interesting =)

File: functions/Get-DbaBackupInformation.ps1 [hunk 1]
```diff
@@ -220,24 +228,38 @@ function Get-DbaBackupInformation {
...
+                            if ($IgnoreLogBackup -and $f -notlike '*LOG' ) {
+                                Write-Message -Level Verbose -Message "Skipping Log backups for Maintenance backups" 
+                            }
+                            else {
+                                $Files += Get-ChildItem -Path $f.fullname -File -Recurse:$DirectoryRecurse
+                            }
                         }
                         elseif ($true -eq $MaintenanceSolution) {
-                            $Files += Get-OlaHRestoreFile -Path $f.fullname -IgnoreLogBackup:$IgnoreLogBackup
+                            $Files += Get-ChildItem -Path $f.fullname
```
Comment (Zhunya) [inline functions/Get-DbaBackupInformation.ps1]: May be add `-Recurse`

**Comment by Stuart-Moore** [comment]
Added recurse to the GCI call @Zhunya  spotted.

**Comment by potatoqualitee** [comment]
@tootblan would you be able to give this a few whirls?

**Comment by tootblan** [comment]
@potatoqualitee What's the best way to test this out? Still learning all the git(hub) stuff. Just download the 5 files in this PR and overwrite them in my local copy of the module?

**Comment by tootblan** [comment]
Tested and I can confirm this fixed my problem. Thanks everybody!

**Comment by potatoqualitee** [comment]
fantastic work, @Stuart-Moore ! will merge into gallery and master momentarily.

thanks so much for testing @tootblan !

**Comment by potatoqualitee** [comment]
merged 👍

**Pull request merged**